python/felis/datamodel.py

```diff
@@ -0,0 +1,354 @@
+            try:
+                ucd.parse_ucd(ivoa_ucd, check_controlled_vocabulary=True, has_colon=";" in ivoa_ucd)
+            except ValueError as e:
+                raise ValueError(f"Invalid IVOA UCD: {e}")
+        return ivoa_ucd
+
+    @model_validator(mode="before")
+    @classmethod
+    def check_units(cls, values: dict[str, Any]) -> dict[str, Any]:
+        """Check that units are valid."""
+        fits_unit = values.get("fits:tunit")
+        ivoa_unit = values.get("ivoa:unit")
+
+        if fits_unit and ivoa_unit:
+            raise ValueError("Column cannot have both FITS and IVOA units")
+        elif fits_unit:
+            unit = fits_unit
+        elif ivoa_unit:
+            unit = ivoa_unit
+        else:
+            unit = None
+
+        if unit is not None:
+            try:
+                units.Unit(unit)
+            except ValueError as e:
+                raise ValueError(f"Invalid unit: {e}")
+
+        return values
+
+
+class Constraint(BaseObject, ABC):
+    """A database table constraint."""
+
+    deferrable: bool | None = False
+    """If `True` then this constraint will be declared as deferrable."""
+
+    initially: str | None = None
+    """Value for ``INITIALLY`` clause, only used if ``deferrable`` is True."""
+
+    annotations: Mapping[str, Any] | None = dict()
+    """Additional annotations for this constraint."""
+
+    type: str | None = Field(None, alias="@type")
+    """The type of the constraint."""
+
+
+class CheckConstraint(Constraint):
+    """A check constraint on a table."""
+
+    expression: str
+
+
+class UniqueConstraint(Constraint):
+    """A unique constraint on a table."""
+
+    columns: list[str]
+
+
+class Index(BaseObject):
+    """A database table index.
+
+    An index can be defined on either columns or expressions, but not both.
+    """
+
+    columns: list[str] | None = None
+    """The columns in the index."""
+
+    expressions: list[str] | None = None
+    """The expressions in the index."""
+
+    @model_validator(mode="before")
+    @classmethod
+    def check_columns_or_expressions(cls, values: dict[str, Any]) -> dict[str, Any]:
+        """Check that columns or expressions are specified, but not both."""
+        if "columns" in values and "expressions" in values:
+            raise ValueError("Defining columns and expressions is not valid")
+        elif "columns" not in values and "expressions" not in values:
+            raise ValueError("Must define columns or expressions")
+        return values
+
+
+class ForeignKeyConstraint(Constraint):
+    """A foreign key constraint on a table."""
+
+    columns: list[str]
+    referenced_columns: list[str] = Field(alias="referencedColumns")
+
+
+class Table(BaseObject):
+    """A database table."""
+
+    columns: list[Column]
+    """The columns in the table."""
+
+    constraints: list[Constraint] | None = []
+    """The constraints on the table."""
+
+    indexes: list[Index] | None = []
```

Comment:
```suggestion
    indexes: list[Index] | None = Field(default_factory=list)
```
Are this and the above ever really None?

Reply:
I have removed the `None` as it is unnecessary.